libyul/interpreter/Scope.cpp

```diff
@@ -0,0 +1,67 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libyul/interpreter/Scope.h>
+
+#include <liblangutil/Exceptions.h>
+
+#include <libyul/AST.h>
+
+using namespace solidity;
+using namespace solidity::yul;
+using namespace solidity::yul::interpreter;
+
+Scope* Scope::getSubscope(Block const& _block)
+{
+	auto [it, isNew] = m_subScopes.try_emplace(&_block, nullptr);
+	if (!isNew)
+		return it->second.get();
+
+	it->second = std::make_unique<Scope>(this);
+	Scope* subscope = it->second.get();
+
+	for (auto const& statement: _block.statements)
+		if (auto const* functionDefinition = std::get_if<FunctionDefinition>(&statement))
+			subscope->m_definedFunctions.emplace(functionDefinition->name, *functionDefinition);
+
+	return subscope;
```

Comment:
This initialization looks like something that belongs in the constructor. Here I'd do just this:
```suggestion
	auto& [block, subScope] = it->second;
	subScope = std::make_unique<Scope>(this, block);
	return subScope.get();
```
and let the scope find its own functions.

Also, I did not expect this function to create a scope since it's called `getSubscope()`. `createSubscope()` would be a more self-explanatory name.

Reply:
I have moved the function assignment to constructor, and renamed the function as well.